packages/wonder-stuff-sentry/src/kind-sentry-error.js

```diff
@@ -0,0 +1,177 @@
+import type {SentryData} from "./types.js";
+
+const EmptySentryData: $ReadOnly<SentryData> = Object.freeze({
+    tags: {},
+    contexts: {},
+    fingerprint: [],
+});
+
+/**
+ * Options for constructing a `KindError`.
+ */
+type Options = {|
+    /**
+     * An error responsible for the error being created.
+     *
+     * @type {?Error}
+     */
+    cause?: ?Error,
+
+    /**
+     * Data to be attached to the error that would be reported to Sentry.
+     *
+     * @type {?$ReadOnly<$Partial<SentryData>>}
+     */
+    sentryData?: ?$ReadOnly<$Partial<SentryData>>,
+
+    /**
+     * Data to be attached to the error that will not get reported to Sentry.
+     *
+     * @type {?$ReadOnly<Metadata>}
+     */
+    metadata?: ?$ReadOnly<Metadata>,
+
+    /**
+     * A prefix to be added to the error name.
+     *
+     * @type {?string}
+     */
+    prefix?: ?string,
+
+    /**
+     * A name for the error.
+     *
+     * @type {?string}
+     */
+    name?: ?string,
+
+    /**
+     * The number of stack frames to strip from the error.
+     *
+     * @type {?number}
+     */
+    stripStackFrames?: ?number,
+
+    /**
+     * The minimum number of stack frames to try and retain.
+     *
+     * @type {?number}
+     */
+    minimumFrameCount?: ?number,
+|};
+
+/**
+ * An error that can be used to capture details for reporting to Sentry.
+ *
+ * The error kind will be a tag on the reported Sentry event.
+ *
+ * @export
+ * @class KindSentryError
+ * @extends {KindError}
+ */
+export class KindSentryError extends KindError {
+    constructor(
+        message: string,
+        kind: string = Errors.Unknown,
+        // It's perfectly valid for options to be an empty object.
+        // $FlowIgnore[unclear-type]
+        options: Options = ({}: any),
+    ) {
+        const {metadata, sentryData, name, ...restOptions} = options;
+        // We want to combine sentry data and metadata into a single object
+        // to send to our super class since folks using this error likely
+        // still want the sentry data to go wherever metadata goes.
+        // We have a couple of options to avoid name collisions:
+        //   1. Assign the given metadata and sentryData to independent keys.
+        //   2. Check that metadata doesn't contain `sentryData` and use that
+        //      key, either overwriting what is there, or erroring if this
+        //      reserved key is used.
+        // Option 1 is cleaner but adds indentation to the data.
+        // Option 2 may make more readable data when stringified, but needs
+        // a policy on what to do about a key collision.
+        // For simplicity of implementation and ease of API use, we choose
+        // option 1.
+        super(message, kind, {
+            ...restOptions,
+            name: name ?? "Sentry",
+            metadata: {
+                sentry: {
+                    // We set the defaults here so that we know these will be
```

Comment:
~SentryData is _next to_ Metadata in the Options arg, but then in here it's added to the `metadata` object. That feels weird to me? Not saying that it needs to change, but just that it feels weird... maybe I just don't understand the code well enough yet...~

OMG OMG OMG, totally missed the above comment 🤦‍♀️

Reply:
I mean, it does still feel weird though... maybe we don't need to add SentryData to metadata? Like, once in-use, if it's a SentryError, it's a SentryError? For typing and stuff too... I dunno, can't explain, but feels like an anti-pattern.

Reply:
Another option is to give the super-class, kind-error a "void-star" like pointer to something arbitrary that all sub-classes can use. It's kind of like exactly what you're doing with `Metadata`, but instead of having `SentryData` as a child of `Metadata`, it just _is_ `Metadata`.

Reply:
There are many cases where we have data we want to log to the console or places but DO NOT want to send to sentry, so in webapp we have this division between those concepts. That's why this is done here too. Some consumers just want all the data, some, like our sentry integration, only want sentry. This achieves that in a controlled manner.